modules/thread-pings/src/main/java/net/neoforged/camelot/module/threadpings/ThreadPingsCommand.java

```diff
@@ -227,32 +294,89 @@ private static CommonResult executeCommon(SlashCommandEvent event) {
+    private static ChannelPingsData fetchPingsData(long channelId, JDA jda) {
+        final var rawPingRoles = BotMain.getModule(ThreadPingsModule.class).db().withExtension(ThreadPingsDAO.class,
+                threadPings -> threadPings.query(channelId));
+        final FilteredRoles pingRoles = filterRoles(rawPingRoles, jda);
+
+        final var rawExemptRoles = BotMain.getModule(ThreadPingsModule.class).db().withExtension(ThreadPingsExemptionsDAO.class,
+                threadPingsExemptions -> threadPingsExemptions.query(channelId));
+        final FilteredRoles exemptRoles = filterRoles(rawExemptRoles, jda);
+
+        return new ChannelPingsData(pingRoles.known(), exemptRoles.known(), pingRoles.unknown(), exemptRoles.unknown());
+    }
+
+    record ChannelPingsData(List<Role> pingRoles, List<Role> exemptRoles, List<Long> unknownPingRoles,
+                            List<Long> unknownExemptRoles) {
+    }
+
+    private static FilteredRoles filterRoles(List<Long> rawRoles, JDA jda) {
+        final var known = new ArrayList<Role>();
+        final var unknown = new ArrayList<Long>();
+        for (long rawRoleId : rawRoles) {
+            final @Nullable Role roleById = jda.getRoleById(rawRoleId);
+            if (roleById != null) {
+                known.add(roleById);
+            } else {
+                unknown.add(rawRoleId);
+            }
+        }
+        return new FilteredRoles(known, unknown);
     }
 
-    record CommonResult(SlashCommandInteraction interaction, GuildChannelUnion channel, List<Role> roles) {
+    record FilteredRoles(List<Role> known, List<Long> unknown) {
     }
 
-    private static String buildMessage(String underText, List<? extends IMentionable> roles) {
+    private static String buildMessage(String place, List<? extends IMentionable> roles, @Nullable List<? extends IMentionable> exemptRoles) {
         final StringBuilder builder = new StringBuilder();
 
-        builder.append("The following roles will be mentioned in public threads created under ")
-                .append(underText)
-                .append(":\n");
+        builder.append("### For __")
+                .append(place)
+                .append("__:")
+                .append('\n');
+
+        builder.append("The following roles will be mentioned in newly-created public threads:")
+                .append('\n');
 
         if (roles.isEmpty()) {
             builder.append("_None._");
         } else {
             builder.append(roles.stream().map(IMentionable::getAsMention).collect(Collectors.joining(", ")));
         }
+        builder.append('\n');
+
+        if (exemptRoles != null) {
+            builder.append('\n')
+                    .append("The following roles are *exempt* from being mentioned in newly-created public threads:")
+                    .append('\n');
+
+            if (exemptRoles.isEmpty()) {
+                builder.append("_None._");
```

Comment:
Not so sure I see the point in this `None`. I would add the not empty check to the not null check.

Reply:
The `configure-channel` subcommand's text will always display the configured ping and exempt roles to the user, to ensure that even if there is a problem with processing the select menu's selected roles, the user is always informed of what the bot has saved for the channel.

Because the `buildMessage` method is used by both `configure-channel` and `configure-guild`, I put in a special-case for `configure-guild` where the exempt roles text is not printed (as it makes no sense in that context) when `exemptRoles` is `null`. If it is a list (empty or note), then it prints the message as it means the method was called for `configured-guild`